### ISSUE TITLE
Implement support for pinvoke for wasm target.

### ISSUE THREAD

**Issue by yowl** [mono/mono#8007]
## Steps to Reproduce

1. See https://github.com/mono/mono/blob/63795c0ce8e564344a791c6406a6862525b7309f/mono/utils/mono-dl-wasm.c#L46.  
2. Its not implemented so calls to another SO using Pinvoke fail.  This would be useful to add Skia support.
3. 

### On which platforms did you notice this

[ ] macOS
[ ] Linux
[ ] Windows
[*] Wasm

**Version Used**:

master

### Stacktrace

```
Please paste the stack trace here if available.
```

**Comment by kumpera** [comment]
Dynamic linking on WebAssembly is very tricky, it requires to compile the runtime with such feature enabled and it comes at a performance & footprint cost. 

I'm not entirely sure on how to proceed with this, but I agree that this is a feature we need to support.

**Comment by kekekeks** [comment]
@kumpera You are [providing a way](https://github.com/mono/mono/tree/master/sdks/wasm) to build a custom `.wasm` file by linking libmonosgen-2.0.a and using a custom driver.c. 

That allows people to statically link additional modules which means that functions from them will be callable using `call_indirect` without additional performance costs. So at least `[DllImport("__Internal")]` should be supported, which would be sufficient for projects like Avalonia that will probably need to ship a custom runtime anyway.

**Comment by kekekeks** [comment]
P/Invoke would also be very nice for calling external JS functions "directly" (yes, I understand that JS side will need manual marshaling) instead of using `eval`.

**Comment by kumpera** [comment]
@kekekeks yeah, the problem with __Internal is that there's no way to locate those symbols in the wasm file as they don't have names.

On using pinvoke to directly call JS, that's even more complicated as the marshaling on the C# side is non trivial.

I explored this problem in the past and the linkage model of WebAssembly is quite unfriendly to dotnet's pinvoke.

I expect this to be solved by a mix of runtime code plus tooling to generate the stubs needed to locate those functions.

**Comment by kumpera** [comment]
@kekekeks makes sense on adding the extra call to register functions, I'll look into it.

When I say i's non trivial to marshal from the C# side is that pinvoking into JS requires currying an extra argument as the only way to call into JS is through the imports table.

Now, on linking shared modules, it requires a lot more than just passing the function pointers around, we need to deal with linear memory linking or you won't be able to pass pointers around.

**Comment by TheSpydog** [comment]
What's the status on this? Are we any closer to being able to DllImport external wasm files or statically link them? I noticed there's a packager option for `--pinvoke-libs` but no documentation or samples demonstrating how to use it. Some clarification here would be greatly appreciated. :)

**Comment by vargaz** [comment]
--pinvoke-libs should be passed the native library filename, i.e. libfoo.a which should be compiled by emscripten.

**Comment by TheSpydog** [comment]
> --pinvoke-libs should be passed the native library filename, i.e. libfoo.a which should be compiled by emscripten.

That's what I thought, but I couldn't get it to work. The library should be compiled as a wasm32 side module, right? What should the dll name string be for DllImport?

**Comment by vargaz** [comment]
It should be compiled as normal wasm object files, not as a side module. The dll name doesn't matter since the pinvoke library will be statically linked into the app.

**Comment by TheSpydog** [comment]
Ah, I see. I can now build and link without any reported issues, but the pinvoke call still fails at runtime with a `DllNotFoundException: __Internal`.

Here's the C# code that causes this exception:
```csharp
using System;
using System.Runtime.InteropServices;

namespace MyNamespace
{
   public class MyProgram
   {
      [DllImport("__Internal", CallingConvention = CallingConvention.Cdecl)
      static extern string SayHello();

      public static void Main(string[] args)
      {
         Console.WriteLine(SayHello());
      }
   }
}
```

The C library:
```c
const char* SayHello()
{
    return "Hello from C library!";
}
```
Compiled with `emcc foo.c -o foo.a`. Linked in the packager tool with `--pinvoke-libs=foo.a`

What am I missing here?

**Comment by vargaz** [comment]
My bad, got confused by our lack of documentation :). The argument to pinvoke-libs is actually the dllimport name, i.e. --pinvoke-libs=__Internal. There is no argument to link a native lib into the final app right now, will add it.

**Comment by vargaz** [comment]
https://github.com/mono/mono/pull/17344

**Comment by Bellt-GCA** [comment]
Would this work for client-side Blazor as well? Basically I would like to call [wbar](https://github.com/chrp/wbar) barcode scanning component from client-side Blazor.  
@TheSpydog could you share your working sample, at this moment I am not able to get something to work...

**Comment by TheSpydog** [comment]
From what I can gather, pinvoke only works with AOT. (I'd love to be proven wrong though!) I'm not sure if it's possible with blazor since I believe it uses the interpreter.

For an AOT app, though, my sample code above should still work, I think. I used [this tutorial](https://balintpogatsa.github.io/2019/05/05/webassembly-mono-aot-example.html) to set up an AOT wasm project. Then in my C# code, I used `[DllImport("foo")]`, specified `--pinvoke-libs=foo --native-lib=foo.a` as packager arguments, and tossed the native library into the folder with build.ninja. This works for simple libraries, at least.
